project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/services/GitWrapperService.kt

```diff
@@ -0,0 +1,127 @@
+
+/**
+ * Service responsible for git operations within ProjectGitSnapshotManager.
+ */
+@Service(Service.Level.APP)
+class GitWrapperService {
+    private val importantFiles = setOf("modules.xml", "misc.xml", "libraries")
+
+    suspend fun commitChanges(context: IJDDContext): IJDDContext = context.apply {
+        commit(this.projectDir)
+        projectDir.refresh(false, true)
+    }
+
+    suspend fun resetChanges(context: IJDDContext) {
+        withContext(Dispatchers.IO) {
+            // git reset --hard
+            Git.open(context.projectDir.toNioPath().toFile()).apply {
+                reset().setMode(ResetCommand.ResetType.HARD).call()
+                clean().setCleanDirectories(true).call()
+                close()
+            }
+        }
+    }
+
+    private suspend fun commit(projectDir: VirtualFile) = withContext(Dispatchers.IO) {
+        Git.open(projectDir.toNioPath().toFile()).let {
+            projectDir.gitAdd(it) { file ->
+                isImportant(file, projectDir)
+            }
+            it.commit().setMessage(UUID.randomUUID().toString()).setAllowEmpty(true)
+                .call()
+            it.close()
+        }
+    }
+
+    private fun isImportant(file: VirtualFile, root: VirtualFile): Boolean {
+        val path = file.toNioPath().relativeTo(root.toNioPath())
+        if (isProjectOrWorkspaceFile(file) && file.name != Project.DIRECTORY_STORE_FOLDER) {
+            val pathString = path.pathString
+            return importantFiles.any { it in pathString }
+        }
+        return true
+    }
+
+    fun gitInitOrOpen(VFProjectDir: VirtualFile): Git {
+        val projectDir: File = VFProjectDir.toNioPath().toFile()
+        if (projectDir.resolve(".git").exists()) {
+            return Git.open(projectDir)
+        }
+        return Git.init()
+            .setDirectory(projectDir)
+            .call()
+            .apply {
+                commit().setMessage("init commit").call()
+            }
+    }
+
+    private fun isCommitListEmpty(git: Git): Boolean = try {
+        git.log().call()  // This will throw NoHeadException if there are no commits
+        false
+    } catch (e: NoHeadException) {
+        true
+    } finally {
+        git.close()
+    }
+
+    /* returns all commits in the chronically reversed order */
+    /* getCommitList(git)[0] == HEAD */
+    public fun getCommitList(git: Git): List<String> {
+        if (isCommitListEmpty(git)) {
+            return listOf()
+        }
+        return git.log()
+            .all()
+            .call()
+            .map { it.name }
+    }
+
+    private suspend fun VirtualFile.gitAdd(git: Git, filter: (VirtualFile) -> Boolean) {
+        if (!filter(this)) {
+            return
+        }
+        val originalPath = this.toNioPathOrNull() ?: return
+        val projectDir = git.repository.directory.parentFile
+        if (originalPath.toString().contains(".git")) {
+            // merge with the filter
+            return
+        }
```

Comment:
I rather hope this situations will never happen, but if the project directory name is `smth.git`, then we're cooked.

For the simple solution you can iterate over the path and check if there any `.git` on the way. 

But I rather recommend using a `git.repository.directory`